evo_researcher/benchmark/benchmark.py

```diff
@@ -25,14 +28,22 @@ def __init__(
         markets: t.List[Market],
         agents: t.List[AbstractBenchmarkedAgent],
         metric_fns: t.Dict[str, t.Callable] = {},
+        cache_path: t.Optional[str] = None,
     ):
         self.markets: t.List[Market] = markets
         self.registered_agents: t.List[AbstractBenchmarkedAgent] = agents
 
         # Predictions
-        self.predictions: t.Dict[str, t.List[PredictionResult]] = {
-            agent.agent_name: [] for agent in self.registered_agents
-        }
+        """
+        TODO also rename PredictionResult -> Prediction
+        """
+        self.cache_path = cache_path
+        if self.cache_path and os.path.exists(self.cache_path):
+            self.predictions = PredictionsCache.load(
```

Comment:
Is there a reason not to use Pydantic's `dict()` method and just save/load the whole Pydantic model? Seems easier than having our own serialization logic, plus `self.predictions` could be typed Pydantic model.

Reply:
How do you mean? We are loading the whole `PredictionsCache` pydantic model (`cls.parse_file(path)`), but we also need custom logic to keep the relevant predictions (and similarly to overwrite the relevant predictions when saving)

Reply:
I meant something like

```python
class PredictionsCache(BaseModel):
  predictions: Predictions

  def add_prediction(self, agent_name: str, question: str, prediction: Any) -> None:
     # Logic as in the agent class.
     ...

  def save(self, path: str): 
    # There shouldn't be a need to re-load `path` content, because we are operating on self, 
    # I'd say we can assume we won't re-write another file that this was loaded from, and if so, it sounds like a bug.
    with open(path, "w") as f:
        json.dump(self.dict(), f, indent=2)
  
  @staticmethod
  def load(path: str) -> "PredictionsCache":
    # Instead of returning dictionary, we can return  `PredictionsCache` itself directly, 
    with open(path, "r") as f:
      return PredictionsCache.parse_obj(json.load(f))
```

I know that now, `PredictionsCache` can hold extra markets, but on the other hand, it's a lot simpler code (it's just a cache, cache can hold a lot of extra stuff)

And in the `benchmark` script, we need to access agent's predictions as predictions_cache[agent][question] anyway, because it's not guaranteed that `predictions_cache[agent].items()` will have the same order as `self.markets`

Reply:
Hmm, but this will just move the filtering logic to the `Benchmarker.get_markets_summary` method

Reply:
If I'm not missing something (and it might be that I am 😄), all we need to do is always iterate over `self.markets` or `self.registered_agents`, instead of iterating over predictions cache. No filtering is needed, or is it?

Reply:
Yeah that's a fair point. Okay, will switch to this!